scheduler.py

```diff
@@ -0,0 +1,443 @@
+from otp4gb import config, cost, otp, parameters, routing
+
+# pylint: enable=wrong-import-position
+
+##### CONSTANTS #####
+
+LOG = logging.getLogger(__name__)
+_CONFIG_FILE = pathlib.Path(__file__).with_suffix(".yml")
+
+_OTP_COST_METRICS_DROP_COLUMNS = [
+    "origin",
+    "destination",
+    "origin_zone_system",
+    "destination_zone_system",
+]
+_OTP_COST_METRICS_RENAME_COLUMNS = {
+    "origin_id": "origin_zones_id",
+    "destination_id": "destination_zones_id",
+}
+_OTP_CENTROIDS_RENAME = {
+    "id": "zone_id",
+    "name": "zone_name",
+    "zone_type_id": "zone_system",
+}
+
+
+def _get_env_bool(name: str, default: bool) -> bool:
+    value = os.environ.get(name, default)
+    if isinstance(value, bool):
+        return value
+
+    value = value.strip().lower()
+    if value in ("yes", "y", "true", "t", "1"):
+        return True
+    if value in ("no", "n", "false", "f", "0"):
+        return False
+
+    raise ValueError(f"invalid boolean value for environment variable {name}={value}")
+
+
+_SHUTDOWN_ENV_VARIABLE = "BUS_ANALYTICS_SHUTDOWN"
+_SHUTDOWN = _get_env_bool(_SHUTDOWN_ENV_VARIABLE, True)
+
+##### CLASSES & FUNCTIONS #####
+
+
+@dataclasses.dataclass
+class OTPParameters:
+    osm_file: str
+    time_periods: list[config.TimePeriod]
+    modes: list[list[routing.Mode]]
+    generalised_cost_factors: cost.GeneralisedCostFactors
+    iterinary_aggregation_method: cost.AggregationMethod = cost.AggregationMethod.MEAN
+    max_walk_distance: int = pydantic.Field(2500, ge=0)
+    number_of_threads: int = pydantic.Field(0, ge=0, le=10)
+    crowfly_max_distance: Optional[float] = None
+    ruc_lookup: Optional[parameters.RUCLookup] = None
+
+
+@dataclasses.dataclass
+class ZoningSystemParams:
+    id: int
+    name: str
+    extents: Optional[config.Bounds] = None
+
+
+class SchedulerConfig(ctk.BaseConfig):
+    output_folder: pydantic.DirectoryPath
+    database_parameters: database.DatabaseConfig
+    otp_parameters: OTPParameters
+    zoning_system_parameters: list[ZoningSystemParams]
+
+
+@dataclasses.dataclass
+class TimetableData:
+    id: int
+    path: pydantic.FilePath
+    date: datetime.date
+
+
+def next_day(date: datetime.date, weekday: int) -> datetime.date:
+    """Get date of next specific weekday."""
+    if weekday > 6 or weekday < 0:
+        raise ValueError(f"invalid value for weekday: {weekday}")
+
+    days = (weekday - date.weekday() + 7) % 7
+    return date + datetime.timedelta(days=days)
+
+
+def download_timetable_to_assets(
+    pg_database: database.Database,
+    adjusted: bool = False,
+    folder: pathlib.Path = config.ASSET_DIR,
+) -> TimetableData:
+    """Get GTFS timetable from database and save to `folder`."""
+    timetable = pg_database.find_recent_timetable(adjusted)
+    if timetable is None:
+        raise ValueError("couldn't find a timetable")
+
+    path = shutil.copy(timetable.actual_timetable_path, folder)
+
+    date = next_day(timetable.upload_date, 0)
```

Comment:
Is there a possibility that we manually upload a historic timetable and the upload_date is out of range of the timetable?

Don't think this is an issue though as I think the timetable downloader should upload the timetable itself so would only be an issue if there was an old timetable manually uploaded which I don't think will happen much if at all.